stackslib/src/chainstate/stacks/boot/signers.clar

```diff
@@ -1,24 +1,43 @@
 (define-data-var last-set-cycle uint u0)
-(define-data-var stackerdb-signer-slots (list 4000 { signer: principal, num-slots: uint }) (list))
+(define-data-var stackerdb-signer-slots-0 (list 4000 { signer: principal, num-slots: uint }) (list))
+(define-data-var stackerdb-signer-slots-1 (list 4000 { signer: principal, num-slots: uint }) (list))
 (define-map cycle-set-height uint uint)
 (define-constant MAX_WRITES u340282366920938463463374607431768211455)
 (define-constant CHUNK_SIZE (* u2 u1024 u1024))
+(define-constant ERR_NO_SUCH_PAGE u1)
+(define-constant ERR_CYCLE_NOT_SET u2)
+(define-map cycle-signer-set uint (list 4000 { signer: principal, weight: uint }))
 
 (define-private (stackerdb-set-signer-slots 
                    (signer-slots (list 4000 { signer: principal, num-slots: uint }))
                    (reward-cycle uint)
                    (set-at-height uint))
-	(begin
+	(let ((cycle-mod (mod reward-cycle u2)))
         (map-set cycle-set-height reward-cycle set-at-height)
         (var-set last-set-cycle reward-cycle)
-		(ok (var-set stackerdb-signer-slots signer-slots))))
+        (if (is-eq cycle-mod u0)
+            (ok (var-set stackerdb-signer-slots-0 signer-slots))
+            (ok (var-set stackerdb-signer-slots-1 signer-slots)))))
 
-(define-read-only (stackerdb-get-signer-slots)
-	(ok (var-get stackerdb-signer-slots)))
+(define-private (stackerdb-set-signers
+                 (reward-cycle uint)
+                 (signers (list 4000 { signer: principal, weight: uint })))
+     (begin
+      (asserts! (is-eq (var-get last-set-cycle) reward-cycle) (err ERR_CYCLE_NOT_SET))
+      (ok (map-set cycle-signer-set reward-cycle signers))))
 
-(define-read-only (stackerdb-get-signer-by-index (signer-index uint))
-	(ok (element-at (var-get stackerdb-signer-slots) signer-index))
-)
+(define-read-only (get-signers (cycle uint))
+     (map-get? cycle-signer-set cycle))
+
+(define-read-only (stackerdb-get-page-count) (ok u2))
+
+(define-read-only (stackerdb-get-signer-slots (page uint))
```

Comment:
The page size limit comes from Clarity itself -- a variable can have type `(list 5139 { signer: principal, num-slots: uint })`, but not `(list 5140 { signer: principal, num-slots: uint } })` because it would permit a value whose representation is too big.  So we need a list size limit.  We can have an unbounded number of pages though.

Reply:
Right— that’s what my latest commit does. It enforces three limits: first on the size of the list, second on the size of the total number of slots across all pages, and third on the number of pages. The .signers contract is implemented with 2 pages.

Reply:
The page limit needs to be a lot bigger, though.  Each signer is getting 12 slots, so at 4000 possible signers, that's 48,000 slots.  So, we need at least 24 pages -- 12 pages for the current signer set, and 12 for the upcoming signer set.

Reply:
The only limits applied to pages are (1) the total inventory size limit, which is set at 2*NUM_SLOTS_PER_USER*4000, and (2) a limit on the _list_ length of 4096. So each page can hold 4000*12 slots just fine.